### DISCUSSION TITLE
When MST node is detached or destroyed?

### DISCUSSION THREAD

**Discussion by baeharam** [mobxjs/mobx-state-tree#1941] [Q&A] [answered]
In official docs, MST is composed of "nodes" which represents `types.model`. So, when I fetch asynchronous data using `flow()` and store response to node, this response is kept until node is destoryed.

Above explanation is what I understood through docs. Is it right? Then, when node is destoryed or detached from "tree"? I'm using this with React but I don't know the lifecycle(NOT LIFECYCLE HOOK) of node. To handle data properly, I need to know when node is destroyed or detached.

**Is node detached or destroyed when I call related method or another situation?** Thanks.

**Comment by EmilTholin** [comment, marked as answer]
Hi @baeharam!

[As the documentation states](https://mobx-state-tree.js.org/overview/hooks), the only thing that will detach the node is an explicit `detach(node)` call. However, a node will be destroyed by a `destroy(node)` call, or by removing or replacing the node from the tree.

**Example**
```js
import { types, flow } from "mobx-state-tree";

const sleep = () => new Promise((resolve) => setTimeout(resolve, 100));

const Model = types.model("Model").actions((self) => ({
  beforeDestroy() {
    console.log("destroy!");
  }
}));

const Store = types
  .model("Store", {
    model: types.maybe(Model)
  })
  .actions((self) => ({
    afterCreate: flow(function* () {
      self.model = Model.create();

      yield sleep();

      self.model = undefined;
    })
  }));

Store.create();
// destroy!
```
